File: 3 course/6 semester/Software development for mobile platforms/films_viewer/lib/Services/ProfileService.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import '../Models/UserProfile.dart';

class ProfileService {
  final FirebaseFirestore _firestore = FirebaseFirestore.instance;

  // Получение профиля пользователя
  Stream<UserProfile> getProfileStream(String userId) {
    return _firestore.collection('users').doc(userId).snapshots().map((snapshot) {
      if (snapshot.exists) {
        return UserProfile.fromFirestore(snapshot.data()!, snapshot.id);
      } else {
        return UserProfile.empty(userId);
      }
    });
  }

  // Обновление профиля
  Future<void> updateProfile(UserProfile profile) async {
    await _firestore
        .collection('users')
        .doc(profile.id)
        .update(profile.toFirestore());
  }

  // Создание профиля при первом входе
  Future<void> createInitialProfile(UserProfile profile) async {
    final doc = await _firestore.collection('users').doc(profile.id).get();
    if (!doc.exists) {
      var dictionary = profile.toFirestore();
      dictionary["emailVerified"] = false;
      dictionary["createdAt"] = FieldValue.serverTimestamp();
      await _firestore.collection('users')
          .doc(profile.id)
          .set(dictionary);
    }
  }
}

File: 3 course/6 semester/Software development for mobile platforms/films_viewer_kotlin/app/src/main/java/com/example/films_viewer_kotlin/services/ProfileService.kt
package com.example.films_viewer_kotlin.services

import com.example.films_viewer_kotlin.models.UserProfile
import com.google.android.gms.tasks.Task
import com.google.android.gms.tasks.Tasks
import com.google.firebase.firestore.FieldValue
import com.google.firebase.firestore.FirebaseFirestore
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.tasks.await

class ProfileService {

    private val firestore = FirebaseFirestore.getInstance()

    fun getProfileStream(userId: String): Flow<UserProfile> = callbackFlow {
        val docRef = firestore.collection("users").document(userId)
        val listenerRegistration = docRef.addSnapshotListener { snapshot, error ->
            if (error != null) {
                close(error)
                return@addSnapshotListener
            }
            if (snapshot != null && snapshot.exists()) {
                val data = snapshot.data
                if (data != null) {
                    val profile = UserProfile.fromFirestore(data, snapshot.id)
                    trySend(profile)
                }
            } else {
                trySend(UserProfile.empty(userId))
            }
        }
        awaitClose { listenerRegistration.remove() }
    }

    suspend fun updateProfile(profile: UserProfile) {
        firestore.collection("users")
            .document(profile.id)
            .update(profile.toFirestore())
            .await()
    }

    fun createInitialProfile(profile: UserProfile): Task<Void> {
        val docRef = firestore.collection("users").document(profile.id)
        return docRef.get().continueWithTask { task ->
            if (!task.result.exists()) {
                val dictionary = profile.toFirestore().toMutableMap()
                dictionary["emailVerified"] = false
                dictionary["createdAt"] = FieldValue.serverTimestamp()
                docRef.set(dictionary)
            } else {
                Tasks.forResult(null)
            }
        }
    }
}
